File: 2026POTD/JANUARY/27-01-2026/3650.java
class Solution {
    public int minCost(int n, int[][] edges) {
        List<int[]>[] g = new ArrayList[n];
        List<int[]>[] in = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            g[i] = new ArrayList<>();
            in[i] = new ArrayList<>();
        }
        for (int[] e : edges) {
            g[e[0]].add(new int[]{e[1], e[2]});
            in[e[1]].add(new int[]{e[0], e[2]});
        }

        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        PriorityQueue<long[]> pq = new PriorityQueue<>(Comparator.comparingLong(a -> a[0]));
        dist[0] = 0;
        pq.offer(new long[]{0, 0});

        while (!pq.isEmpty()) {
            long[] cur = pq.poll();
            long d = cur[0];
            int u = (int) cur[1];
            if (d != dist[u]) continue;
            if (u == n - 1) return (int) d;

            for (int[] e : g[u]) {
                if (dist[e[0]] > d + e[1]) {
                    dist[e[0]] = d + e[1];
                    pq.offer(new long[]{dist[e[0]], e[0]});
                }
            }
            for (int[] e : in[u]) {
                if (dist[e[0]] > d + 2L * e[1]) {
                    dist[e[0]] = d + 2L * e[1];
                    pq.offer(new long[]{dist[e[0]], e[0]});
                }
            }
        }
        return -1;
    }
}

File: 2026POTD/JANUARY/27-01-2026/3650.cpp
class Solution {
public:
    int minCost(int n, vector<vector<int>>& edges) {
        vector<vector<pair<int,int>>> g(n), in(n);
        for (auto &e : edges) {
            g[e[0]].push_back({e[1], e[2]});
            in[e[1]].push_back({e[0], e[2]});
        }

        const long long INF = 1e18;
        vector<long long> dist(n, INF);
        priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<>> pq;

        dist[0] = 0;
        pq.push({0, 0});

        while (!pq.empty()) {
            auto [d, u] = pq.top(); pq.pop();
            if (d != dist[u]) continue;
            if (u == n - 1) return d;

            for (auto &[v, w] : g[u]) {
                if (dist[v] > d + w) {
                    dist[v] = d + w;
                    pq.push({dist[v], v});
                }
            }

            for (auto &[v, w] : in[u]) {
                if (dist[v] > d + 2LL * w) {
                    dist[v] = d + 2LL * w;
                    pq.push({dist[v], v});
                }
            }
        }

        return -1;
    }
};